src/unix/linux-core.c

```diff
@@ -519,19 +535,20 @@ static int read_models(unsigned int numcpus, uv_cpu_info_t* ci) {
 }
 
 
-static int read_times(FILE* statfile_fp,
-                      unsigned int numcpus,
-                      uv_cpu_info_t* ci) {
+/* Return numcpus read (always <= numcpus input as argument). */
```

Comment:
If I'm not getting wrong with it, it is returning the new `numcpus` the `read_times` function has read.

Reply:
Yes. Why is that a problem?

Reply:
Sorry I didn't get the question clear. The `read_times` may read fewer cpus on race conditions of CPU offline/online switches. It reads at most the number as the input parameter `numcpus`, and returns the number it actually reads.

Would you mind clarifying what you would like to see?

Reply:
I don't see the motivation for the retry loop, and not instead returning the smaller of the `numcpus` read

Reply:
It is because we already allocated a space for `numcpus` to be read, and filled the space with `read_models`. We can not assure which CPU is offline, thus the already filled data may not be aligned with the data read by `read_times`. Retry with realloc is the straightforward way to re-align the data.

Reply:
You don't know what CPU is online anyways, since that is a TOCTOU problem you cannot solve.

Reply:
Or could we fill the times data with a default value, like 0? If returning a number less than what we actually allocated, we'd either need to release the trailing ones before returning from `uv_cpu_info()`, or it would be a problem on `uv_free_cpu_info(uv_cpu_info_t* cpu_infos, int count)`.

Reply:
If I take a couple CPUs offline, it turns out that the numbering of CPUs appears to generally be constant (and not consecutive), so we could figure out whether the CPU stat matches the CPU info from that info:

```
$ cat /proc/stat
cpu  68001332 23837 3965504 2967709685 686128 0 3590631 0 0 0
cpu0 2831878 6270 561933 756678005 189422 0 1707110 0 0 0
cpu1 60664155 5876 2514339 697226786 165419 0 137158 0 0 0
cpu2 2679201 5648 451204 757166559 182552 0 146598 0 0 0
cpu3 1818383 5470 433948 755025194 148436 0 1598168 0 0 0
cpu6 2150 0 1232 6820 0 0 31 0 0 0
```

Reply:
not necessarily, but it does suggest we can be more accurate about matching up the online cpus than just checking the count

Reply:
If we allocate a `uv_cpu_info_s` array lengthed 5 for the above example, how can we save the information of cpu6 in the `uv_cpu_info_s` array?

We don't have a cpuN mark in the structure of `uv_cpu_info_s`. I may not clearly get your point of "matching up the online cpus than just checking the count" part. Changing the structure by adding a cpuN mark to it may be ABI breaking and I didn't find it is worth trying.

Reply:
We could store the information only internally here between functions perhaps, or restructure the code to iterate both files together? Not sure

Reply:
This sounds like we need to check and align the `cpuN`s from different reads and different files. And we may need to filter out the `cpuN`s that is missing from one file but present in another, in the final `uv_cpu_info_s` result. Sounds like a bunch of refactoring.

As @bnoordhuis has approved the PR already, @bnoordhuis would you mind sharing your idea on this topic? thank you!

Reply:
I'm not wholly clear on what the discussion is about but w.r.t. the retry loop: a CPU coming online/going offline is an uncommon occurrence so retrying once or twice seems okay because chances are excellent the next attempt succeeds.

Reply:
@vtjnash what do you think to decrease the retry times as suggested by Ben? Does that address your concern about the retry loop?

Reply:
@bnoordhuis @vtjnash updated with the max_retry reduced to 2. Would you mind taking a look again?

Reply:
friendly ping @bnoordhuis @vtjnash

Reply:
friendly ping @bnoordhuis @vtjnash